### ISSUE TITLE
gdbserver doesn't work

### ISSUE THREAD

**Issue by floren** [Harvey-OS/harvey#464]
Debugging is important, and being able to use gdb is one of the benefits of switching to clang/gcc; if we can no longer use acid, we need something.

#### Expected behavior
Should be able to remote debug processes

#### Actual behavior
gdbserver exits and does nothing

#### Steps to reproduce
- Find pid of program you want to debug
- run gdbserver -p <pid>
- on LInux box, run "gdb <binary>"
- at gdb prompt, type "target remote <vm ip>:1666"

**Comment by rminnich** [comment]
John, this used to work. See what alvaro says but he claimed it sort of still did. But, yet, in my view this is THE HIGHEST PRIORITY item for harvey, much higher than any other work including Apex, because if you can't debug you're wasting time. At this point we've wasted several person-years because we lack this simple capability.

**Comment by fuchicar** [comment]
I have a different behaviour (I'm using Debian Stable and maybe the gdb versions are very different). I connect to gdbserver, but when it sends "stop" command to the process to be debugged, the write call never ends, so client give me a timeout.

I tryed to do it manually (echo stop > /proc/<pid>/ctl) and I obtained the same result: It never returned the prompt

**Comment by rminnich** [comment]
This is a good start but you need to experiment a bit more. 

So do this:
start a dd

Pretend the dd has pid 1234
in another window, do this
echo stop > /proc/1234/ctl

Note the echo hangs. Go back to the dd window, and hit return.

note the stop completes and dd is stopped. The issue here seems to be that since dd is blocked on input, hence never scheduled, we're blocking until it is run. Is this correct behavior? I don't know.

**Comment by rminnich** [comment]
So my next suggestion: take a loop at sys/src/9/port/devproc.c, look for "stop" in quotes, and work out how that command "stop" gets managed, and what happens in devwrite to make the process stop. 

Look at the function that manages it. Why would that function block? Is it really a problem? Is gdb being unreasonable to wait? Is your test reasonable?

**Comment by fuchicar** [comment]
Ok, I see. It's in the middle of a system call "read", only when the system call ends the process can be stopped.

Sorry for the noise. Thanks a lot Ron for you patient with me.

**Comment by rminnich** [comment]
It's great you are doing this. We need to build our collective knowledge. Don't forget, we all start from zero on this software and have to learn.

**Comment by floren** [comment]
Testing procedure:

In one rio window, run dd
Get the pid of that dd. Let’s pretend it’s 314159
In another window:

    ratrace -c gdbserver -p 314159

On the linux host:

    gdb amd64/bin/dd 

Once in gdb:

    set architecture i386:x86-64
    target remote harvey-host:1666

In the dd window, hit return
And watch gdbserver crash and burn …

**Comment by gmacd** [comment]
This particular issue should be fixed (PR #511) now that we don't connect, and we only pass one pid.  Further problems can be described in newer issues.

**Comment by floren** [comment]
This is great, @gmacd 

I tested it on a broken samterm (another issue i should file) and was able to get this:

(gdb) target remote 192.168.0.2:1666
Remote debugging using 192.168.0.2:1666
0x0000000000425b60 in _threadassert (s=0x43e798 "npcs < 6") at debug.c:59
59              abort();
(gdb) bt
#0  0x0000000000425b60 in _threadassert (s=0x43e798 "npcs < 6") at debug.c:59
#1  0x000000000042e18c in _assert (s=0x43e798 "npcs < 6") at port/_assert.c:19
#2  0x0000000000437682 in getcallstack (pcs=0x662700, npcs=6) at port/getcallstack.c:17
#3  0x000000000042fa4f in ppanic (p=0x64d9f8 <sbrkmem>, fmt=0x43e374 "pool panic") at port/malloc.c:174
#4  0x00000000004323ef in panicblock (p=0x64d9f8 <sbrkmem>, b=0x67df88, msg=0x43e253 "corrupt tail magic0") at port/pool.c:752
#5  0x0000000000432103 in blockcheck (p=0x64d9f8 <sbrkmem>, b=0x67df88) at port/pool.c:811
#6  0x0000000000433ee6 in arenamerge (p=0x64d9f8 <sbrkmem>, bot=0x65deb0, top=0x67e440) at port/pool.c:698
#7  0x0000000000433cd4 in poolnewarena (p=0x64d9f8 <sbrkmem>, asize=4128) at port/pool.c:643
#8  0x0000000000430785 in poolallocl (p=0x64d9f8 <sbrkmem>, dsize=112) at port/pool.c:981
#9  0x00000000004304f2 in poolalloc (p=0x64d9f8 <sbrkmem>, n=112) at port/pool.c:1223
#10 0x000000000042f0b7 in malloc (size=48) at port/malloc.c:212
#11 0x0000000000426417 in _threadmalloc (size=48, z=1) at lib.c:22
#12 0x0000000000423af4 in chancreate (elemsize=4, elemcnt=0) at channel.c:83
#13 0x0000000000406bc7 in duplicate (l=0x406bc7 <duplicate+103>, r=..., f=0x662a60, close=11) at main.c:203
#14 0x000000000040a7ea in hsetsnarf (nc=1) at mesg.c:645
#15 0x0000000000400115 in iconinit () at icons.c:60
#16 0x00000000004267e7 in mainlauncher (arg=0x65df18) at main.c:71
#17 0x0000000000427e8a in launcheramd64 (f=0x4267c0 <mainlauncher>, arg=0x65df18) at amd64.c:19
#18 0xfefefefefefefefe in ?? ()
#19 0xfefefefefefefefe in ?? ()
#20 0x00000000004267c0 in ?? () at main.c:179
#21 0x000000000065df18 in ?? ()
#22 0x00000000faf0f1fe in ?? ()
#23 0x0000000000000000 in ?? ()

I'm closing this issue, because as of now GDB *does* work. Let's file issues for enhancements henceforth.
